packages/solid-plotly.js/src/factory.tsx

```diff
@@ -55,220 +55,101 @@ const updateEvents: PlotlyHTMLElementEventName[] = [
+const isBasicEvent = (eventName: PlotlyHTMLElementEventName): eventName is PlotlyBasicEvent => {
+  return updateEvents.includes(eventName)
+}
+
 const isBrowser = typeof window !== 'undefined'
 
 export default function plotComponentFactory(Plotly: typeof PlotlyInstance) {
   const PlotlyComponent = (props: PlotlyComponentProps): JSX.Element => {
-    const [plotEl, setPlotEl] = createSignal<PlotlyHTMLElementWithListener | null>(null)
-    let resizeHandler: (() => void) | null = null
-    const handlers: Partial<Record<PlotlyHTMLElementEventName, EventHandler>> = {}
-    const data = () => props.data
-    const layout = () => props.layout
-    const config = () => props.config
-    const frames = () => props.frames
-    const revision = () => props.revision
-
-    const attachUpdateEvents = () => {
-      const el = plotEl()
-      if (!el || typeof el.on !== 'function') return
-      updateEvents.forEach(updateEvent => {
-        el.on(updateEvent as PlotlyBasicEvent, handleUpdate)
-      })
-    }
-
-    const removeUpdateEvents = () => {
-      const el = plotEl()
-      if (!el || typeof el.removeListener !== 'function') return
-      updateEvents.forEach(updateEvent => {
-        el.removeListener(updateEvent as PlotlyBasicEvent, handleUpdate)
-      })
-    }
-
-    const handleUpdate = () => {
-      figureCallback(props.onUpdate)
-    }
-
-    const figureCallback = (callback?: EventHandler) => {
-      const el = plotEl()
-      if (typeof callback === 'function' && el) {
-        const figure: PlotlyFigure = {
-          data: el.data as Partial<PlotData>[],
-          layout: el.layout,
-          frames: el._transitionData ? el._transitionData._frames : undefined,
-        }
-        callback(figure, el)
+    const onRef = (el: PlotlyHTMLElementWithListener) => {
+      const initUpdateEvents = () => {
+        if (typeof el.on !== 'function') return
+
+        const handleUpdate = () => props.onUpdate?.(getCurrentFigure(), el)
+        updateEvents.forEach(updateEvent => {
+          el.on(updateEvent as PlotlyBasicEvent, handleUpdate)
+        })
+
+        onCleanup(() => {
+          updateEvents.forEach(updateEvent => {
+            el.removeListener(updateEvent as PlotlyBasicEvent, handleUpdate)
+          })
+        })
       }
-    }
 
-    const syncWindowResize = (invoke: boolean) => {
-      const el = plotEl()
-
-      if (!isBrowser) return
-
-      if (props.useResizeHandler && !resizeHandler) {
-        resizeHandler = () => el && Plotly.Plots.resize(el)
-        window.addEventListener('resize', resizeHandler)
-        if (invoke) {
-          resizeHandler()
-        }
-      } else if (!props.useResizeHandler && resizeHandler) {
-        window.removeEventListener('resize', resizeHandler)
-        resizeHandler = null
+      const initEventHandlers = () => {
+        Object.entries(eventNameMapping).forEach(([propName, plotlyEventName]) => {
+          createEffect(() => {
```

Comment:
Since there are no signals in this `createEffect` function that I can see will it every be invoked beyond the first time? 

I don't think it will which means the only purpose of the `createEffect` is to have the `onCleanup` closure created.

Reply:
This references a prop `const handler = props[propName as keyof PlotlyEventHandlers]` but not a signal. Maybe that creates auto tracking for the prop?

Reply:
> This references a prop const handler = props[propName as keyof PlotlyEventHandlers] but not a signal. Maybe that creates auto tracking for the prop?

Yes exactly. Props are transformed to getters: `<Component prop={signal()}/>` you can visualise as `Component({ get prop() { return signal() })`, so `props[propName]` looks like a simple property access, but is actually a signal call.

This makes that whenever you access a prop in an effect, it will re-run that effect whenever the prop changes. this is also the reason why you should not destructure props in solid, because then you would not call the signal in the effect, but it would be called when destructuring.